### DISCUSSION TITLE
How many percent is tailwind v4 ready and how to track it?

### DISCUSSION THREAD

**Discussion by MrOxMasTer** [tailwindlabs/tailwindcss#13938] [General]
It's been about 4 months and it's interesting. How much has already been done since the beginning of the announcement of the alpha version, how much more will be needed and will all the chips announced in the roadmap be implemented?

**Comment by adamwathan** [comment]
Hey! We haven't done much work on it since the alpha was release as we had to prioritize a bunch of work on [Catalyst](https://tailwindui.com/templates/catalyst) to get it out of development preview, but now that that's out of the way we're getting back into v4 work this week so you'll start to see a lot more progress happening soon.

Hoping to fill in all of the backwards compatibility holes and move from alpha to beta by the end of the summer, then tag a stable release later in the year once the community has had enough of a chance to help us identify and work out any of the kinks during the beta period.

**Reply by HiggsWRX** [reply]
Hi Adam,

Not the OP but thanks for the information, very exciting to read! In my project we chose to adopt talwind v4 already since it's such a breeze in setup and configuration when compared to 3.4.

One quick question, do you have an idea when theming via selector strategy will be reintroduced back? More towards the end of the alpha/beta phase? Or anywhere soon?

 Our project is huge and while not really that important, we're waiting for it to be back on v4 to introduce light/dark theme support so we only have to set it up in one place instead of literally every class having to have a `dark:` counterpart.

Thanks for the time!

**Reply by sceee** [reply]
Hi Adam,

Thanks for your answer, despite v4 still being alpha, it seems to be pretty solid already and nice to work with. 👏 

In addition to the questions from the others I'm wondering if you plan to have any large/major breaking changes coming to v4 beta/stable or whether these further updates will mostly be additions/fixes?
Essentially my question is: if v4 fits my needs at the moment (which it does and which is why I would use it already in production) - is there any substantial reason to hold back and wait for the beta/stable as an update could suddenly break my whole application and require a major workover (e.g. not just adjusting some lines)?
I know this can always happen but according to what I've read I have the feeling the existing "feature set" will most likely stay at least the same and there are just some fixes/missing features from v3 that will be added towards the stable release.

Any insights would be appreciated, thanks!

**Reply by MrOxMasTer** [reply]
> Thanks for your answer, despite v4 still being alpha, it seems to be pretty solid already and nice to work with. 👏

So I'm asking about it. About the job. Have they returned at least all the basic functions that existed in tailwind v3.4, for example important and plugins and support tailwind.config.js. Because I would also like to use v4 already, but I just don't understand if the basic functionality works and that's why I created this discussion to find out how many percent of the roadmap is ready, because I was surprised that there is support for basic functions that were already in v3.4

**Reply by MrOxMasTer** [reply]
> Hoping to fill in all of the backwards compatibility holes and move from alpha to beta by the end of the summer, then tag a stable release later in the year once the community has had enough of a chance to help us identify and work out any of the kinks during the beta period.

will there be any PR on this roadmap?
![image](https://github.com/user-attachments/assets/8d5099ac-4bec-4c39-b8fb-02ceeae5cade)

Like this:
![image](https://github.com/user-attachments/assets/4f582b81-9dea-45fa-845e-ae7a97b0717a)
![image](https://github.com/user-attachments/assets/57ca65cc-e156-4489-b15a-0f2972b8d04e)

To understand if at least the old functionality has been returned

**Reply by MrOxMasTer** [reply]
> Hoping to fill in all of the backwards compatibility holes and move from alpha to beta by the end of the summer, then tag a stable release later in the year once the community has had enough of a chance to help us identify and work out any of the kinks during the beta period.

?

**Reply by adamwathan** [reply]
https://github.com/tailwindlabs/tailwindcss/blob/next/CHANGELOG.md

**Reply by mendozagit** [reply]
Hey guys, QQ, there is any date in mind to release v4 version? Thanks!

**Reply by adamwathan** [reply]
This month 💪🏻

**Comment by zoltanszogyenyi** [comment]
Hey everyone!

Now that there's a v4-beta available, I've made an [upgrade with a Tailwind v3 project to v4-beta](https://flowbite.com/blog/tailwind-v4/) and the process was fairly simple, however here are a few feedbacks from my part:

- initially it was a bit of a blockage that the form and button styles changed, I think the upgrade process will be easier now that the change has been reverted
- all `!important` tags in the previous CSS file will throw an error to the compiler, do we want this? Adding the `!` prefix to each class is kind of cumbersome
- it's a bit interesting to me that the fontSize variable returns something like `[size, { lineheight: x }]` - this might throw an error for the current Tailwind forms plugin too
- the rounded and shadow classes which are documented, that wasn't too hard to change
- it's a bit limitative that when using `addComponents` or `addBase` I can't have another function that basically adds simple styles without using the Tailwind `@layer` directives, I might be wrong in assuming the general architecture though

Anyways, the upgrade is really nice and I think the the automatic content paths algorithm is a good addition, the speed also and getting closer to the native CSS layers and variables. Keep up the great job!

**Comment by MrOxMasTer** [comment]
It remains to add a function, which I have already implemented:
https://github.com/MrOxMasTer/tailwindcss-multiple-classes
and will be generally good

**Comment by DanielOberlechner** [comment]
Hi @all,
I would like to ask if there are any updates for the release date of TailwindCSS v4?

Is there a time window when we can expect it?
I would like to know it if I should port a current project to version 3 or already be prepared to work with version 4?

Kind regards,
Daniel

**Reply by zoltanszogyenyi** [reply]
It will be launched this week based on Adam's latest tweet.

**Reply by mendozagit** [reply]
![image](https://github.com/user-attachments/assets/42ba71eb-7b31-4c3f-860a-7606a3e65f11)

**Reply by KlotzJesse** [reply]
![image](https://github.com/user-attachments/assets/4b5c2fc2-718a-4b9c-bdbe-cab54e6d74a3)
4.0.0 got released with "next" tag 12hrs ago ;)

**Comment by MrOxMasTer** [comment]
someone is waiting for a new TV show or movie, and we're waiting for a new version of the library to come out. Ahhhh
